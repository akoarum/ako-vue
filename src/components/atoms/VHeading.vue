<script lang="ts">
import Vue, { VNode, CreateElement } from 'vue'

export default Vue.extend({
  props: {
    level: { type: Number, default: 2 },
    visual: { type: Number, default: 2 },
    underline: { type: Boolean }
  },
  computed: {
    styledClasses(): string[] {
      const classes: string[] = []

      classes.push('vHeading')
      classes.push(`-lv${this.visual}`)
      if (this.underline) classes.push('-underline')

      return classes
    }
  },
  render(createElement: CreateElement): VNode {
    return createElement(
      `h${this.level}`,
      { class: this.styledClasses.join(' ') },
      this.$slots.default
    )
  }
})
</script>

<style lang="scss" scoped>
.-lv2 {
  @include media-min {
    font-size: 2.8rem;
    line-height: 1.3;

    &.-underline {
      padding: 0 3px 6px;
    }
  }

  @include media-max {
    font-size: 2.4rem;
    line-height: 1.4;

    &.-underline {
      padding: 0 3px 4px;
    }
  }
}

.-lv3 {
  @include media-min {
    font-size: 2.5rem;
    line-height: 1.4;

    &.-underline {
      padding: 0 3px 4px;
    }
  }

  @include media-max {
    font-size: 2.2rem;
    line-height: 1.4;

    &.-underline {
      padding: 0 3px 4px;
    }
  }
}

.-lv4 {
  @include media-min {
    font-size: 2.2rem;
    line-height: 1.4;

    &.-underline {
      padding: 0 3px 4px;
    }
  }

  @include media-max {
    font-size: 2rem;
    line-height: 1.4;

    &.-underline {
      padding: 0 3px 3px;
    }
  }
}

.-underline {
  border-bottom: 1px solid #ccc;
}
</style>
