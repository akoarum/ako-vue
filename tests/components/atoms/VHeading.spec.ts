import { shallowMount } from '@vue/test-utils'
import VHeading from '~/components/atoms/VHeading.vue'

describe('VHeading', () => {
  let wrapper: any

  beforeEach(() => {
    wrapper = shallowMount(VHeading, {
      propsData: {},
      slots: {
        default: '見出しテキスト'
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('rendering', () => {
    it('levelに応じてタグが変わる', () => {
      expect(wrapper.is('h2')).toBe(true)

      wrapper.setProps({ level: 4 })
      expect(wrapper.is('h4')).toBe(true)
    })
  })

  describe('computed', () => {
    it('styledClasses: propsを基にクラス配列を生成して返す', () => {
      expect(wrapper.vm.styledClasses).toEqual(['vHeading', '-lv2'])

      wrapper.setProps({ visual: 3 })
      expect(wrapper.vm.styledClasses).toEqual(['vHeading', '-lv3'])

      wrapper.setProps({ underline: true })
      expect(wrapper.vm.styledClasses).toEqual(['vHeading', '-lv3', '-underline'])
    })
  })
})
