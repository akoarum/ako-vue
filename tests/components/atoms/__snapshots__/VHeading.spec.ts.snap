// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VHeading initialized snapshot 1`] = `
<h2
  class="vHeading -lv2"
>
  見出しテキスト
</h2>
`;
